crowdfunding_compassion/models/crowdfunding_project.py

```diff
@@ -220,35 +220,34 @@ def validate(self):
             )
 
     @api.model
-    def get_active_projects(self, limit=None, year=None):
-        if year:
-            self = self.search(
-                [
-                    ("deadline", ">=", datetime(year, 1, 1)),
-                    ("deadline", "<=", datetime(year, 12, 31)),
-                ]
-            )
-        # get active projects, from most urgent to least urgent
+    def get_active_projects(self, limit=None, year=None, type=None):
+        # Get active projects, from most urgent to least urgent
         active_projects = self.search(
             [
-                ("state", "!=", "draft"),
                 ("deadline", ">=", date.today()),
             ], limit=limit, order="deadline ASC"
         )
+
+        # Get finished projects, from most recent to oldest expiring date
         finished_projects = self.env[self._name]
         if not limit or (limit and len(active_projects) < limit):
-            # get finished projects, from most recent to oldest expiring date
-            finish_limit = None
-            if limit:
-                finish_limit = limit-len(active_projects)
+            finish_limit = limit-len(active_projects) if limit else None
             finished_projects = self.search(
                 [
-                    ("state", "!=", "draft"),
                     ("deadline", "<", date.today()),
                 ], limit=finish_limit, order="deadline DESC"
             )
 
-        return active_projects + finished_projects
+        projects = active_projects + finished_projects
+
+        filters = [
+            ("state", "!=", "draft"),
+            ("deadline", ">=", datetime(year, 1, 1)) if year else None,
+            ("deadline", "<=", datetime(year, 12, 31)) if year else None,
+            ("type", "=", type) if type else None
+        ]
+
+        return projects.search(list(filter(None, filters)))
```

Comment:
I think the `search` method acts on all records, this line would be equivalent of `self.env['crowdfunding.project'].search([..])`, meaning that it could return projects that were not present in the records at the beginning. That's why I suggest to start by filtering before doing the join of finished and active projects.

Reply:
Oh I see, that's kind of an odd behavior. I updated the function and applied the filters in the same search() call as for the dates.